File: python/services/repository/ports/git_client.py
"""GitClient port — git remote operations (A.3)."""
from __future__ import annotations

from typing import Protocol, runtime_checkable

from services.repository.domain.domain import Branch, ConnectionStatus


@runtime_checkable
class GitClient(Protocol):
    """Driven port for git remote operations. NoOpGitClient implements this (TODO)."""

    async def test_connection(
        self, remote_url: str, credential_ref: str
    ) -> ConnectionStatus:
        """Probe remoteURL using credentialRef and return the connection status."""
        ...

    async def list_branches(
        self, remote_url: str, credential_ref: str
    ) -> list[Branch]:
        """Return the branches available on the remote."""
        ...

    async def push_result(
        self,
        remote_url: str,
        credential_ref: str,
        target_branch: str,
        create_new_repo: bool,
    ) -> tuple[str, str]:
        """Push migration output.  Returns (pushed_branch, new_repo_url)."""
        ...

File: core/services/repository/ports/git_client.go
package ports

import (
	"context"

	"milton_prism/core/services/repository/domain"
)

// GitClient is the driven port for git remote operations.
type GitClient interface {
	// ProbeSource checks remoteURL without cloning. It determines whether the
	// URL is reachable, its visibility (public/private), and whether the
	// supplied token grants read access. An empty token is valid for public repos.
	ProbeSource(ctx context.Context, remoteURL, token string) (*domain.SourceProbeResult, error)

	// TestConnection probes remoteURL using credentialRef and returns the
	// connection status. It does not update the repository record; the
	// application layer is responsible for persisting the result.
	TestConnection(ctx context.Context, remoteURL, credentialRef string) (domain.ConnectionStatus, error)

	// ListBranches returns the branches available on the remote.
	ListBranches(ctx context.Context, remoteURL, credentialRef string) ([]*domain.Branch, error)

	// PreflightTarget validates a push destination WITHOUT pushing anything.
	// It probes the git smart-HTTP receive-pack (write) endpoint to determine
	// reachability and whether writeToken grants push access, and lists the
	// remote refs to determine whether the target repository is empty (A.3).
	// No commit, no ref update, no clone is performed. writeToken is supplied as
	// HTTP Basic Auth only — never logged or embedded in any URL or error message.
	PreflightTarget(ctx context.Context, targetURL, writeToken string) (*domain.TargetPreflightResult, error)

	// PushResult initializes a temporary workspace, writes files preserving
	// directory structure, commits them with commitMessage (a default traceable
	// message is used when empty), and pushes to targetURL.
	// writeToken is supplied as HTTP Basic Auth only — never logged or embedded
	// in any URL or error message.
	PushResult(ctx context.Context, targetURL, writeToken string, files []*domain.PushFile, commitMessage string) (pushedBranch string, err error)
}
